src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,536 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+
+namespace Microsoft.ML.Runtime.Internal.CpuMath
+{
+    internal static class SseIntrinsics
+    {
+        private static unsafe Vector128<float> Load1(float* src, int* idx)
+        {
+            return Sse.SetScalarVector128(src[idx[0]]);
+        }
+
+        private static unsafe Vector128<float> Load4(float* src, int* idx)
+        {
+            return Sse.SetVector128(src[idx[3]], src[idx[2]], src[idx[1]], src[idx[0]]);
+        }
+
+        private static Vector128<float> Rotate(Vector128<float> x)
+        {
+            return Sse.Shuffle(x, x, 0x39);
+        }
+
+        private static Vector128<float> RotateReverse(Vector128<float> x)
+        {
+            return Sse.Shuffle(x, x, 0x93);
+        }
+
+        // Warning: this operation changes the value of x => do not reuse x
+        private static unsafe void Store4(Vector128<float> x, float* dst, int* idx)
+        {
+            Sse.StoreScalar(dst + idx[0], x);
+            for (int i = 1; i <= 3; i++)
+            {
+                x = Rotate(x);
+                Sse.StoreScalar(dst + idx[i], x);
+            }
+        }
+
+        internal static void ScaleU(float scale, float[] dst, int count)
+        {
+            Vector128<float> scaleVector = Sse.SetAllVector128(scale);
+
+            unsafe
+            {
+                fixed (float* pdst = dst)
+                {
+                    float* pDstCurrent = pdst;
+                    float* pEnd = pdst + count;
+
+                    while (pDstCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
+
+                        dstVector = Sse.Multiply(scaleVector, dstVector);
+                        Sse.Store(pDstCurrent, dstVector);
+
+                        pDstCurrent += 4;
+                    }
+
+                    while (pDstCurrent < pEnd)
```

Comment:
Generally speaking, you can finish off the remaining indices one of two ways.

Like this, where you drop down to a scalar algorithm or, you can "double compute" a couple of the indices, that involves:
* Saving the stored result (`dstVector`) from the last iteration of the vectorized code
* Moving `pDstCurrent` back such that `pDstCurrent + elementsPerIteration == pEnd`
* Doing a single iteration for the remaining elements
* Mix the saved result from the last iteration of the vectorized code with the result from the remaining elements
* Write the result

This generally results in more performant code, depending on the exact algorithm and number of remaining elements

Reply:
Thank you!  This is much needed.  Same as above: Should I save it for a future follow-up, or should I propagate this fix to other functions at this point?

Reply:
This is probably good for future followup.